Python/Product/Analysis/Interpreter/Ast/AstAnalysisWalker.cs

```diff
@@ -0,0 +1,260 @@
+            if (expr is LambdaExpression) {
+                return _interpreter.GetBuiltinType(BuiltinTypeId.Function);
+            }
+
+            return null;
+        }
+
+        private IMember GetInScope(string name) {
+            foreach (var m in _scope) {
+                if (m == null) {
+                    continue;
+                }
+                IMember obj;
+                if (m.TryGetValue("__class__", out obj)) {
+                    return obj;
+                }
+            }
+            return null;
+        }
+
+        private IPythonType CurrentClass {
+            get {
+                var m = _scope.Peek();
+                if (m != null) {
+                    IMember cls;
+                    m.TryGetValue("__class__", out cls);
+                    return cls as IPythonType;
+                }
+                return null;
+            }
+        }
+
+        public override bool Walk(AssignmentStatement node) {
+            var m = _scope.Peek();
+            if (m != null) {
+                var value = GetValueFromExpression(node.Right, m);
+                foreach (var ne in node.Left.OfType<NameExpression>()) {
+                    m[ne.Name] = value;
+                }
+            }
+            return base.Walk(node);
+        }
+
+        public override bool Walk(ImportStatement node) {
+            var m = _scope.Peek();
+            if (m != null && node.Names != null) {
+                try {
+                    for (int i = 0; i < node.Names.Count; ++i) {
+                        var n = node.AsNames?[i] ?? node.Names[i].Names[0];
+                        if (n != null) {
+                            m[n.Name] = new AstPythonConstant(
+                                _interpreter.GetBuiltinType(BuiltinTypeId.Module),
+                                GetLoc(node.AsNames[i])
+                            );
+                        }
+                    }
+                } catch (IndexOutOfRangeException) {
+                }
+            }
+            return false;
+        }
+
+        public override bool Walk(FromImportStatement node) {
+            var m = _scope.Peek();
+            if (node.Root.MakeString() == "__future__") {
+                return false;
+            }
+
+            if (m != null && node.Names != null) {
+                try {
+                    for (int i = 0; i < node.Names.Count; ++i) {
+                        var n = node.AsNames?[i] ?? node.Names[i];
+                        if (n != null) {
+                            m[n.Name] = new AstPythonConstant(
+                                _interpreter.GetBuiltinType(BuiltinTypeId.Module),
+                                GetLoc(n)
+                            );
+                        }
+                    }
+                } catch (IndexOutOfRangeException) {
+                }
+            }
+            return false;
+        }
+
+        public override bool Walk(FunctionDefinition node) {
```

Comment:
We don't walk into functions though. Any forward references we may encounter would cause SyntaxError at compile time. (When we come to handling annotations it'll be more interesting, as those will require a second pass, but this level of analysis isn't even going to try and determine return values from the body.)

---
In reply to: [83940287](https://github.com/Microsoft/PTVS/pull/1729#discussion_r83940287) [](ancestors = 83940287)